src/Symfony/Component/Security/Core/Encryption/PhpseclibEncryption.php

```diff
@@ -0,0 +1,160 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Encryption;
+
+use phpseclib\Crypt\AES;
+use phpseclib\Crypt\RSA;
+use Symfony\Component\Security\Core\Exception\EncryptionException;
+use Symfony\Component\Security\Core\Exception\InvalidArgumentException;
+use Symfony\Component\Security\Core\Exception\MalformedCipherException;
+use Symfony\Component\Security\Core\Exception\UnsupportedAlgorithmException;
+use Symfony\Component\Security\Core\Exception\WrongEncryptionKeyException;
+
+if (!class_exists(RSA::class)) {
+    throw new \LogicException('You cannot use "Symfony\Component\Security\Core\Encryption\PhpseclibEncryption" as the "phpseclib/phpseclib:2.x" package is not installed. Try running "composer require phpseclib/phpseclib:^2".');
+}
+
+/**
+ * The secret key length should be 32 bytes, but other sizes are accepted.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ * @experimental in 5.2
+ */
+class PhpseclibEncryption implements SymmetricEncryptionInterface, AsymmetricEncryptionInterface
+{
+    /**
+     * @var string application secret
+     */
+    private $secret;
+
+    public function __construct(string $secret)
+    {
+        $this->secret = $secret;
+    }
+
+    public function generateKeypair(): array
+    {
+        $rsa = new RSA();
+        $key = $rsa->createKey();
+
+        if ($key['partialkey']) {
+            throw new EncryptionException('Failed to generate RSA keypair.');
+        }
+
+        return [
+            'public' => $key['publickey'],
+            'private' => $key['privatekey'],
+        ];
+    }
+
+    public function encrypt(string $message, ?string $publicKey = null, ?string $privateKey = null): string
+    {
+        set_error_handler(__CLASS__.'::throwError');
+        $nonce = random_bytes(16);
+        try {
+            if (null === $publicKey) {
+                $algorithm = 'aes';
+                $aes = new AES();
+                $aes->setKey($this->secret);
```

Comment:
Hmm this class is Asymmetric when calling `encrypt` with a `publicKey` arguments, in that case, it does not use the compulsory argument `secret`. Something looks fragile and complicated here.

I'd rather
```
class PhpseclibSymmetricEncryption implement SymmetricEncryptionInterface
class PhpseclibAsymmetricEncryption implement AsymmetricEncryptionInterface
```
or something similar to https://github.com/paragonie/paseto/tree/master/src/Keys
```
class PhpseclibEncryption implement EncryptionInterface
{
    function __construct(KeyInterface $key) {
    } 
}

interface SymmetricKeyInterface extends KeyInterface
interface AsymmetricKeyInterface extends KeyInterface
```

Reply:
These is not "secret" in RSA. I use AES only for symmetric encryption, RSA is for asymmetric.